R/g_mlm.R

```diff
@@ -94,23 +113,29 @@ g_mlm <- function(mod, p_const, mod_denom = mod, r_const = NULL, infotype = "exp
     stop("g_mlm() only available for lme or gls models. Please specify such a model in the 'mod_denom' argument.")
   }
 
-  theta <- extract_varcomp(mod_denom)              # full theta vector
+  theta <- extract_varcomp(mod_denom)
+  if (!is.null(mod$call$weights) && inherits(mod$modelStruct$varStruct, "varIdent") && separate_variances) {
+    theta <- extract_varcomp(mod_denom, separate_variances = TRUE)
+  }
 
   if (is.null(r_const)) {
     warning("The r_const argument was not specified. Defaulting to r_const equal to all 1's. Are you sure this is right?")
     r_const <- rep(1L, length(unlist(theta)))
   }
 
-  r_theta <- sum(unlist(theta) * r_const)                                 # r'theta (sum of var comp)
+  r_theta <- sum(unlist(theta) * r_const)                         # r'theta (sum of var comp)
   delta_AB <- p_beta / sqrt(r_theta)                              # delta_AB
   kappa_sq <- sum(tcrossprod(p_const) * vcov(mod)) / r_theta      # kappa^2
   cnvg_warn <- !is.null(attr(mod_denom,"warning"))                # indicator that RML estimation has not converged
 
   # calculate inverse Fisher information
   info_inv <- varcomp_vcov(mod_denom, type = infotype)
-  SE_theta <- sqrt(diag(info_inv))                                # SE of theta
+  if (!is.null(mod$call$weights) && inherits(mod$modelStruct$varStruct, "varIdent") && separate_variances) {
```

Comment:
Same question: Is there a reason we need to call `extract_varcomp()` twice? Would it work to just pass `separate_variances = separate_variances`?

Reply:
Thanks for catching this.